### PR TITLE
Allow config-file overrides in ecosystem checks

### PR DESCRIPTION
**Pull request by zanieb**
Adds the ability to override `ruff.toml` or `pyproject.toml` settings per-project during ecosystem checks.

Exploring this as a fix for the `setuptools` project error. 

Also useful for including Jupyter Notebooks in the ecosystem checks, see #9293

Note the remaining `sphinx` project error is resolved in #9294

### CODE REVIEW

File: python/ruff-ecosystem/ruff_ecosystem/check.py [hunk 1]
```diff
@@ -479,23 +479,24 @@ async def compare_check(
...
-        )
-        comparison_task = tg.create_task(
-            ruff_check(
-                executable=ruff_comparison_executable.resolve(),
-                path=cloned_repo.path,
-                name=cloned_repo.fullname,
-                options=options,
-            ),
-        )
+    with options.update_toml(cloned_repo.path):
```
Comment (zanieb) [inline python/ruff-ecosystem/ruff_ecosystem/check.py:482]: Technically should be async but I don't mind blocking the event loop to read/write from the file in this simple code.

Needs to be implemented in `format.py` too.

File: python/ruff-ecosystem/ruff_ecosystem/projects.py [hunk 2]
```diff
@@ -52,6 +59,54 @@ def with_options(self: Self, **kwargs) -> Self:
...
+                        continue
+                    else:
+                        key = key[len("preview.") :]
+
+                # Allow "no-preview" only overrides
+                if key.startswith("no-preview."):
+                    if self.preview:
+                        continue
+                    else:
+                        key = key[len("no-preview.") :]
```
Comment (zanieb) [inline python/ruff-ecosystem/ruff_ecosystem/projects.py:91]: This seems a little confusing but something like this is needed to solve the problem.

File: python/ruff-ecosystem/ruff_ecosystem/projects.py [hunk 3]
```diff
@@ -26,14 +31,115 @@ class Project(Serializable):
...
+            base = ["tool", "ruff"]
+
+        overrides = {
+            **self.always,
+            **(self.when_preview if preview else self.when_no_preview),
+        }
+
+        if not overrides:
+            yield
+            return
```
Comment (MichaReiser) [inline python/ruff-ecosystem/ruff_ecosystem/projects.py:110]: I'm sure this is correct, but it reads strangely. Why do we need to `yield` (nothing) right before returning?

**Review by MichaReiser — APPROVED** [review]

File: python/ruff-ecosystem/ruff_ecosystem/projects.py (hunk 3 above)
Comment (zanieb) [reply python/ruff-ecosystem/ruff_ecosystem/projects.py:110]: If we don't yield, the function is no longer a generator and the caller will complain.

```python
from contextlib import contextmanager


@contextmanager
def foo():
    return


with foo():
    pass
```

```
Traceback (most recent call last):
  File "/Users/mz/eng/src/astral-sh/ruff/example.py", line 9, in <module>
    with foo():
  File "/Users/mz/.pyenv/versions/3.12.0/lib/python3.12/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not an iterator
```

**Pull request merged**